cloud/amazon/ec2_ami.py

```diff
@@ -297,17 +448,29 @@ def deregister_image(module, ec2):
     wait = module.params.get('wait')
     wait_timeout = int(module.params.get('wait_timeout'))
 
-    img = ec2.get_image(image_id)
-    if img == None:
-        module.fail_json(msg = "Image %s does not exist" % image_id, changed=False)
-
     try:
-        params = {'image_id': image_id,
-                  'delete_snapshot': delete_snapshot}
-
-        res = ec2.deregister_image(**params)
-    except boto.exception.BotoServerError, e:
-        module.fail_json(msg = "%s: %s" % (e.error_code, e.error_message))
+        img = ec2.get_image(image_id)
+    except boto.exception.EC2ResponseError, e:
+        if 'InvalidAMIID.NotFound' in e.error_code:
+            module.fail_json(msg = "Image %s does not exist" % image_id, changed=False)
+
+    # Get all snapshot ids associated with image
```

Comment:
Getting snapshots is slow and expensive, why not hold off on running this unless you're sure they've passed in an option to delete snapshots?